Enable Google Cloud IAP authentication through spring security

elefeint opened the pull request:
This PR changes/creates 3 projects.

**1. spring-cloud-gcp-autoconfigure.**
New package o.s.c.g.a.security  containing customizations to autoconfigured OAuth Resource Server.

Uses Spring Security OAuth2 Resource Server to parse the IAP header, validate its signature and claims, and provide an authorization object.

There is no additional logic, but two IAP-specific beans are provided through autoconfiguration:
* A version of JwtDecoder that goes against IAP public key and validates the private key using ES256 algorithm. The first part could have been configured with a property, but encryption is not yet property-configurable. When spring-projects/spring-boot#15115 is resolved, this bean can be removed and replaced with the two properties.
* A BearerTokenResolver that parses `x-goog-iap-jwt-assertion `header instead of the standard OAuth `access_token`.

There is one unrelated change in this PR -- bringing spring-security as a transitive dependency made GcpStorageAutoConfigurationTests needlessly secure, so I excluded security autoconfiguration from the test.

**2. spring-cloud-gcp-starters/spring-cloud-gcp-starter-security-iap**
A new project that brings in autoconfiguration and a dependency on `spring-boot-starter-oauth2-resource-server.` 

**3. spring-cloud-gcp-samples/spring-cloud-gcp-security-iap-sample**
A new project containing a minimal demo app showing how to configure a secured URL with a custom `WebSecurityConfigurerAdapter`.

Fixes #1016.

artembilan commented:
Travis report:
```
[ERROR] Failed to execute goal on project spring-cloud-gcp-security-iap-sample: Could not resolve dependencies for project com.example:spring-cloud-gcp-security-iap-sample:jar:1.1.0.BUILD-SNAPSHOT: Could not find artifact org.springframework.cloud:spring-cloud-gcp-starter-security-iap:jar:1.1.0.BUILD-SNAPSHOT in sonatype-snapshots (https://oss.sonatype.org/content/repositories/snapshots/)
```
Missed something in poms? As a module for example?

artembilan commented on spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/security/IapAuthenticationAutoConfiguration.java:
```diff
@@ -0,0 +1,76 @@
+ * @since 1.1
+ */
+@Configuration
+@ConditionalOnProperty("spring.cloud.gcp.security.iap.enabled")
```
Why no `matchIfMissing = true`  as we have in all other auto-configuration in this project?

This property also has to be added to the `additional-spring-configuration-metadata.json`, if you are not going to extract a fully blown `@ConfigurationProperties` class, as you mention with TODO below.

artembilan replied:
I don't see that you have addressed this one.

Do you see my review at all?

Thanks

elefeint replied:
So, this one is interesting. It's true that our normal practice is to have` matchIfMissing = true`, but our normal practice is to, at the same time, add @ConditionalOnClass that will invalidate the autoconfig whenever a required dependency is missing. IAP, however, no longer has any unique required dependencies -- the only dependency it needs is `spring-boot-starter-oauth2-resource-server`, which could be present on the classpath for unrelated reasons (for example, an application uses OAuth and also brings in `spring-cloud-gcp-autoconfigure` with a datastore starter).

I considered adding matchIfMissing and having the starter turn on the property in its application.properties, but that does not seem to be done anywhere else in the project, either. Do you think this is a good alternative?

artembilan replied:
OK. Fine with me.

Since we really don't have a strict dependency and you are definitely right with some side-effect, so let it it as `false` by default!

Then we need explain in the Docs that this feature must be turned on explicitly.
On the other hand is there a choice when we really know that we are in GCP, so that is fully OK to have this IAP enabled by default? Are we really going to use some other JWT mechanism there ?

Thanks for explanation!

elefeint replied:
I could see someone hosting an application on GCP but using Facebook or Twitter for authentication. IAP is more of an enterprise authN/authZ mechanism, since someone needs to set up users and groups for it to be useful.

I am writing documentation now; apparently that takes me even longer than writing code :\

artembilan commented on spring-cloud-gcp-samples/spring-cloud-gcp-security-iap-sample/README.adoc:
```diff
@@ -0,0 +1,7 @@
+TODO
```
TODO?

elefeint replied:
I've been retelling this story in English for years. Finally, I can cite it "in the original Klingon" (well, except when I first encountered it, the difficulty was with getting a painting past government censors, and not a strict editor): https://koppel.pro/diary/sobacka-favorskogo-2443

artembilan commented on spring-cloud-gcp-samples/spring-cloud-gcp-security-iap-sample/README.adoc:
```diff
@@ -0,0 +1,7 @@
+
+Note: this sample should be tested when deployed to AppEngine Flexible or Kubernetes environments.
+
+It is possible, in principle, to grab a recent JWK token from a deployed apps' /headers path, and to test locally, but this token will expire, causing authentication to be declined locally.
```
I believe `/headers` should be wrapped to the code.

artembilan commented on spring-cloud-gcp-samples/spring-cloud-gcp-security-iap-sample/README.adoc:
```diff
@@ -0,0 +1,7 @@
+
+It is possible, in principle, to grab a recent JWK token from a deployed apps' /headers path, and to test locally, but this token will expire, causing authentication to be declined locally.
+
+If you don't create your own implementation of WebSecurityConfigurerAdapter, Spring Boot will introduce its own, OAuth2ResourceServerWebSecurityConfiguration
```
The classes are definitely code. so, please, proper wrapping.
Also every sentence should finish with period.

artembilan commented on spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/security/IapAuthenticationAutoConfiguration.java:
```diff
@@ -54,13 +56,14 @@
-	private static final String PUBLIC_KEY_VERIFICATION_LINK = "https://www.gstatic.com/iap/verify/public_key-jwk";
-
-	private static final String ENCRYPTION_ALGORITHM = "ES256";
+	@VisibleForTesting
```
What is the problem to make all these constants `public`?
to be honest I'm fully against such a `@VisibleForTesting` hack in the production code.
This is not Spring way...

elefeint replied:
Makes sense. It's a cultural difference; I am still adjusting -- in a giant monorepo, a convention to restrict visibility, unless there is a good reason to relax it, is reasonable. But in Spring, where every project's boundaries are clearly delineated, it makes sense to do the opposite.

elefeint commented:
@artembilan Yep, I've seen it, thank you! I haven't finished addressing the feedback yet. I pushed code because I wanted to see how codecov would react to tests.

artembilan commented on spring-cloud-gcp-docs/src/main/asciidoc/security-iap.adoc:
```diff
@@ -0,0 +1,41 @@
+
+https://cloud.google.com/iap/[Cloud Identity-Aware Proxy (IAP)] provides a security layer over applications deployed to Google Cloud.
+
+The IAP starter uses
```
"One sentence per line": https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line

meltsufin commented on spring-cloud-gcp-autoconfigure/pom.xml:
```diff
@@ -193,6 +193,13 @@
+        <!-- Security -->
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-oauth2-resource-server</artifactId>
```
It's odd to see a starter imported here. This module is just a pom. Can we be more specific about which modules we need here? Is it `spring-security-oauth2-resource-server` and something else?

artembilan replied:
Ah! Right. We must not use starters in the `autocofigure` module as dependency.
I don't remember "Why?", but that is what @snicoll recommended us to do.

meltsufin commented on spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/security/IapAuthenticationAutoConfiguration.java:
```diff
@@ -0,0 +1,74 @@
+@ConditionalOnProperty("spring.cloud.gcp.security.iap.enabled")
+@ConditionalOnClass({NimbusJwtDecoderJwkSupport.class})
+@AutoConfigureBefore(OAuth2ResourceServerAutoConfiguration.class)
+public class IapAuthenticationAutoConfiguration {
```
Does it make sense to add and enable `IapAuthenticationProperties`?
It could expose the configuration you currently have via constants below, and the "enabled" property.

artembilan replied:
We have an issue on the matter:  https://github.com/spring-cloud/spring-cloud-gcp/issues/1192

meltsufin commented on spring-cloud-gcp-docs/src/main/asciidoc/security-iap.adoc:
```diff
@@ -0,0 +1,40 @@
+
+https://cloud.google.com/iap/[Cloud Identity-Aware Proxy (IAP)] provides a security layer over applications deployed to Google Cloud.
+
+The IAP starter uses {spring-security-ref}#oauth2resourceserver[Spring Security OAuth 2.0 Resource Server] functionality to automatically extract user identity from the proxy-injected `x-goog-iap-jwt-assertion` header.
```
```suggestion
The IAP starter uses {spring-security-ref}#oauth2resourceserver[Spring Security OAuth 2.0 Resource Server] functionality to automatically extract user identity from the proxy-injected `x-goog-iap-jwt-assertion` HTTP header.
```

meltsufin commented on spring-cloud-gcp-docs/src/main/asciidoc/security-iap.adoc:
```diff
@@ -0,0 +1,40 @@
+
+The IAP starter uses {spring-security-ref}#oauth2resourceserver[Spring Security OAuth 2.0 Resource Server] functionality to automatically extract user identity from the proxy-injected `x-goog-iap-jwt-assertion` header.
+
+To use IAP authentication, you need to:
```
I think you have to use the abbreviation "Cloud IAP" everywhere. We've been dinged for this before when we tried to refer to Cloud Spanner as just "Spanner".
Notice the abbreviation use in [official docs](https://cloud.google.com/iap/).
```suggestion
To use Cloud IAP authentication, you need to:
```

meltsufin commented on spring-cloud-gcp-starters/spring-cloud-gcp-starter-security-iap/pom.xml:
```diff
@@ -0,0 +1,27 @@
+
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-oauth2-resource-server</artifactId>
```
See my comment in the autoconfigure module. This would match whatever we have there as optional deps for this.

artembilan replied:
I think refer to starters from our own starters is OK. 
See `spring-cloud-gcp-starter-sql-mysql` and `spring-cloud-gcp-starter-trace`.
Plus we can borrow some pattern from Spring Boot per se: https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-starters/spring-boot-starter-websocket/pom.xml

artembilan commented on spring-cloud-gcp-starters/spring-cloud-gcp-starter-security-iap/pom.xml:
```diff
@@ -20,8 +20,16 @@
         <dependency>
-            <groupId>org.springframework.boot</groupId>
-            <artifactId>spring-boot-starter-oauth2-resource-server</artifactId>
+            <groupId>org.springframework.security</groupId>
```
Huh? We have just discussed that it is OK to use starters as dependencies in starters.

elefeint replied:
Might as well be consistent with autoconfig. And, I guess, for someone trying to troubleshoot why the starter works but their application does not, it would be nice to see all the dependencies together and not have to do a recursive hunt for them.

artembilan replied:
OK. Let it be. I just wanted to follow the practice from Spring Boot.
Won't hart to have ti like this...

elefeint commented:
@meltsufin 
About GCP-specific claims -- we discussed whether to enforce the container-specific (AppEngine Flex vs Kubernetes) audience claim a couple of weeks ago, and the conclusion was that there was not a ton of value, given that the verification would need to be different depending on where the app is deployed. But we can reevaluate -- building in custom validators would be as simple as setting them while creating `NimbusJwtDecoderJwkSupport` in `IapAuthenticationAutoConfiguration`. If we choose to do it, I would do it in a different PR.

As far as USER_ROLE, this was necessitated by the specific details source (`PreAuthenticatedGrantedAuthoritiesUserDetailsService`) I chose in the previous implementation. Spring security resource server implementation uses a standard `WebAuthenticationDetailsSource` which does not require authorities claims.

wyaeld commented:
maybe not right place to ask, but @elefeint @meltsufin are you able to share when next milestone (or RC) which might include this will be?  Reason I ask is I'm leading an effort to get apps up on GCP and IAP is one of the very next things on my roadmap.  This repo isn't using the Milestone functionality much so its hard to know what the 1.1.0 GA and M3 roadmap looks like.

meltsufin commented:
@wyaeld We are following the Spring Cloud Greenwich release train now. So, [this is the tentative schedule](https://github.com/spring-cloud/spring-cloud-release/milestones) that would correspond to our 1.1.0 releases. Next milestone might be tomorrow if all goes well.
